src/storage-client/src/controller.rs

```diff
@@ -1743,170 +1808,74 @@ mod persist_read_handles {
         tx: UnboundedSender<(tracing::Span, PersistWorkerCmd<T>)>,
     }
 
-    impl<T> Drop for PersistWorker<T>
-    where
-        T: Timestamp + Lattice + Codec64,
-    {
-        fn drop(&mut self) {
-            self.send(PersistWorkerCmd::Shutdown);
-            // TODO: Can't easily block on shutdown occurring.
-        }
-    }
-
     /// Commands for [PersistWorker].
     #[derive(Debug)]
     enum PersistWorkerCmd<T: Timestamp + Lattice + Codec64> {
-        Register(GlobalId, ReadHandle<SourceData, (), T, Diff>),
+        Register(GlobalId, SinceHandle<SourceData, (), T, Diff, i64>),
         Downgrade(BTreeMap<GlobalId, Antichain<T>>),
-        Snapshot(
-            GlobalId,
-            Antichain<T>,
-            tokio::sync::oneshot::Sender<Result<Vec<(Row, Diff)>, StorageError>>,
-        ),
-        Shutdown,
     }
 
     impl<T: Timestamp + Lattice + Codec64> PersistWorker<T> {
         pub(crate) fn new() -> Self {
             let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<(tracing::Span, _)>();
 
-            mz_ore::task::spawn(|| "PersistReadHandles", async move {
-                let mut read_handles: BTreeMap<
-                    GlobalId,
-                    Option<ReadHandle<SourceData, (), T, Diff>>,
-                > = BTreeMap::new();
+            mz_ore::task::spawn(|| "PersistWorker", async move {
+                let mut since_handles = BTreeMap::new();
 
-                let antichain_minimum_element = Antichain::from_elem(T::minimum());
-                let mut shutdown = false;
-                let mut interval = tokio::time::interval(std::time::Duration::from_secs(60));
-                while !shutdown {
-                    tokio::select! {
-                        _ = interval.tick() => {
-                            let futs = FuturesUnordered::new();
-                            for (_id, read) in read_handles.iter_mut() {
-                                if let Some(read) = read.as_mut() {
-                                    // downgrade_since is idempotent, so downgrading to the minimum
-                                    // element is a no-op and can be used as a lease liveness
-                                    // heartbeat.
-                                    //
-                                    // TODO: This behavior might be too subtle. Consider
-                                    // reintroducing a public maybe_heartbeat_reader that internally
-                                    // does this.
-                                    futs.push(read.maybe_downgrade_since(&antichain_minimum_element));
+                while let Some(cmd) = rx.recv().await {
+                    // Peel off all available commands.
+                    // This allows us to catch up if we fall behind on downgrade commands.
+                    let mut commands = vec![cmd];
+                    while let Ok(cmd) = rx.try_recv() {
+                        commands.push(cmd);
+                    }
```

Comment:
very nice! Though, this could, in extremely rare circumstances, starve the code below right?

Reply:
It's hard to imagine this actually happening but yeah. It's the same logic as before though, I just moved the code around

Reply:
yeah we can leave this for now! At some point we may want to add metrics on this channel!